tests/uefi_sb/test_varstored_sb.py

```diff
@@ -153,3 +156,102 @@ def test_key_exchanges(self, uefi_vm):
+        db = EFIAuth("db", other_certs=[ms_certs.db_uefi_2011(), ms_certs.db_win_2011()])
+        # Some test VMs don't like an empty dbx when their own dbx is empty, so just put whatever in there
+        dbx = EFIAuth.self_signed("dbx")
+
+        PK.sign_auth(PK)
+        PK.sign_auth(KEK)
+        KEK.sign_auth(db)
+        KEK.sign_auth(dbx)
+
+        vm.install_uefi_certs([PK, KEK, db, dbx])
+        return [PK, KEK, db, dbx]
+
+    def install_new_certs(self, vm: VM, signer: EFIAuth):
+        """Populate a key set that looks like the new defaults with 2023 MS keys."""
+
+        newPK = EFIAuth.self_signed("PK")
+        newKEK = EFIAuth("KEK", other_certs=[ms_certs.kek_ms_2011(), ms_certs.kek_ms_2023()])
+        newdb = EFIAuth(
+            "db",
+            other_certs=[
+                ms_certs.db_win_2011(),
+                ms_certs.db_win_2023(),
+                ms_certs.db_uefi_2011(),
+                ms_certs.db_uefi_2023(),
+                ms_certs.db_oprom_2023(),
+            ],
+        )
+        newdbx = EFIAuth("dbx")
+
+        newPK.sign_auth(newPK)
+        # Technically, there's no need to sign the other databases since we're setting them from Dom0.
+        # If signing with the old PK works, there'd be no need to test signing with the new PK.
+        # We use an invalid signer to test scenarios where the user mixes and matches default and custom keys.
+        signer.sign_auth(newKEK)
+        signer.sign_auth(newdb)
+        signer.sign_auth(newdbx)
+
+        vm.install_uefi_certs([newPK, newKEK, newdb, newdbx])
+
+    def test_key_upgrade(self, uefi_vm: VM):
+        vm = uefi_vm
+        vm.param_set("platform", True, key="secureboot")
+        assert not vm.get_vtpm_uuid()
+        vm.create_vtpm()
+
+        PK, _, _, _ = self.install_old_certs(vm)
+        boot_and_check_sb_succeeded(vm)
+
+        vm.shutdown(verify=True)
+
+        self.install_new_certs(vm, PK)
+        boot_and_check_sb_succeeded(vm)
+
+    def test_key_upgrade_bitlocker(self, uefi_vm: VM):
+        vm = uefi_vm
+        vm.param_set("platform", True, key="secureboot")
+        assert not vm.get_vtpm_uuid()
+        vm.create_vtpm()
+
+        PK, _, _, _ = self.install_old_certs(vm)
+        boot_and_check_sb_succeeded(vm)
+
+        vm.execute_powershell_script("Add-WindowsFeature BitLocker,EnhancedStorage")
+        vm.reboot(verify=True)
+
+        vm.execute_powershell_script("Enable-BitLocker $Env:SystemDrive -TpmProtector -UsedSpaceOnly")
+        # Confirm if PCR7 is bound.
+        assert vm.execute_powershell_script(
+            r"""Get-CimInstance -Namespace Root\CIMV2\Security\MicrosoftVolumeEncryption `
+-Query "select * from Win32_EncryptableVolume where VolumeType=0" |
+Invoke-CimMethod -MethodName GetSecureBootBindingState |
+Where-Object ReturnValue -eq 0 |
+Select-Object -ExpandProperty BindingState"""
+        ) == "3"  # Bound
+        vm.execute_powershell_script("Suspend-BitLocker $Env:SystemDrive")
```

Comment:
What happens if we don't suspend bitlocker first? Would a test where we don't suspend it be worth it?

Reply:
The BitLocker theatrics here are really just for testing Windows's PCR7 binding logic against PK changes. The suspend is required to unbind PCR7 and prepare for `install_new_certs`. Without it the test would simply fail every time.